docs/source/architecture.rst

```diff
@@ -73,60 +84,40 @@ We recursively parse each of the elements, using their in built grammars.
-  reimplement the sections that they need to. One reason for the *Ref*
+- Grammars are contained in *dialects*, the root dialect being
+  the *ansi* dialect. The ansi dialect is used to host logic common
+  to all dialects, and so does not necessary adhere to the formal
+  ansi specification. Other SQL dialects inherit from the ansi dialect,
+  replacing or patching any segments they need to. One reason for the *Ref*
   grammar is that it allows name resolution of grammar elements at runtime
   and so a *patched* grammar with some elements overriden can still rely on
-  lower level elements which haven't been overwritten.
-- *Segments* and *Grammars* operate roughly interchangeably from a dialect.
-  A grammar is something which can be matched against and return a result.
-  These will be used as instantiated classes and usually contain a set of
-  elements (stored in the *_elements* attribute) which are the sub elements
-  which the grammar will use for matching. These may in turn be a mixture
-  of segments and grammars. A segment can be used as either a class or an
-  instance of a class. While matching, the class itself is used, but if
-  it does match, rather than just returning the original segments it was
-  passed unchanged, it will optionally return mutated versions of those
-  segments which may have been given a more specific meaning by the
-  matcher. For example a *RawSegment* containing *123.4* may actually
-  be mutated to an instance of the *NumericLiteralSegment* class when
-  matched against the *NumericLiteralSegment* class.
+  lower level elements which haven't been redeclared within the dialect
 - All grammars and segments attempt to match as much as they can and will
   return partial matches where possible. It is up to the calling grammar
   or segment to decide whether a partial or complete match is required
   based on the context it is matching in.
 
+
 Stage 4, the linter
 -------------------
 
-Given the complete parse tree, we now walk that tree to assess the tree
-for linting errors. A linter is an object which is able to traverse the
-tree itself, allowing it to choose which objects it examines and which
-it alerts as a problem.
+Given the complete parse tree, rule classes assess for linting errors by
+traversing the tree, looking for segments and patterns of concern. If
+the rule discovers a violation, it returns a :code:`LintResult` pointing
+to the segment which has caused the violation.
 
-Some linters, may optionally be able to *fix* the problems they find. If
-this is the case, they will optionally return a mutated tree as one of
-their return values, which can be passed to the next linter. In normal
-operation this will not be what is returned, because it becomes confusing
-with line references for a user fixing issues manually.
+Some rules are able to *fix* the problems they find. If
+this is the case, they will return a mutated segment, containing
```

Comment:
This doesn't seem quite accurate. Rules can return a list of `fixes`, which describe mutations (i.e. changes) to be made to the tree. This can include edits, inserts, or deletions to the tree.

Reply:
Quite right - thanks!